File: api/app/services/classifier.py
"""Classificação heurística Canal/Filme/Série+Episódio/Não Classificado
(FR-008; research.md R5; ADR-005 §2 — sem evidência, nunca inventar).

Agrupamento de episódios sob a série correta é expresso aqui por
`series_key`/`series_name`; a resolução para `parent_id` real (linkando ao
CatalogItem da série) é responsabilidade do importer (T024), que tem acesso
à sessão do banco.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from app.models.catalog_item import CatalogItemKind
from app.services.m3u_parser import ParsedEntry

_EPISODE_PATTERN = re.compile(
    r"^(?P<base>.*?)[\s._-]*S(?P<season>\d{1,2})\s*E(?P<episode>\d{1,3})\b.*$",
    re.IGNORECASE,
)

_MOVIE_KEYWORDS = {"filme", "filmes", "movie", "movies", "vod"}
_SERIES_KEYWORDS = {"série", "séries", "serie", "series", "seriado"}
_CHANNEL_KEYWORDS = {"canal", "canais", "channel", "channels", "ao vivo", "live", "tv"}


@dataclass(frozen=True)
class ClassifiedEntry:
    kind: CatalogItemKind
    name: str
    original_name: str
    group: str | None
    url: str
    series_key: str | None = None
    series_name: str | None = None
    season_number: int | None = None
    episode_number: int | None = None


def _group_keyword_kind(group: str | None) -> CatalogItemKind | None:
    if not group:
        return None
    lowered = group.lower()
    if any(keyword in lowered for keyword in _MOVIE_KEYWORDS):
        return CatalogItemKind.MOVIE
    if any(keyword in lowered for keyword in _SERIES_KEYWORDS):
        return CatalogItemKind.SERIES
    if any(keyword in lowered for keyword in _CHANNEL_KEYWORDS):
        return CatalogItemKind.CHANNEL
    return None


def classify_entry(entry: ParsedEntry) -> ClassifiedEntry:
    episode_match = _EPISODE_PATTERN.match(entry.name.strip())
    if episode_match:
        base = episode_match.group("base").strip(" -._") or entry.name
        return ClassifiedEntry(
            kind=CatalogItemKind.EPISODE,
            name=entry.name,
            original_name=entry.name,
            group=entry.group,
            url=entry.url,
            series_key=base.lower(),
            series_name=base,
            season_number=int(episode_match.group("season")),
            episode_number=int(episode_match.group("episode")),
        )

    group_kind = _group_keyword_kind(entry.group)
    if group_kind is not None:
        return ClassifiedEntry(
            kind=group_kind,
            name=entry.name,
            original_name=entry.name,
            group=entry.group,
            url=entry.url,
        )

    # Evidência insuficiente — Não Classificado, sem hierarquia inventada.
    return ClassifiedEntry(
        kind=CatalogItemKind.UNCLASSIFIED,
        name=entry.name,
        original_name=entry.name,
        group=entry.group,
        url=entry.url,
    )


def classify_entries(entries: list[ParsedEntry]) -> list[ClassifiedEntry]:
    return [classify_entry(entry) for entry in entries]

File: tv-web/src/lib/catalog/classifier.ts
/**
 * Porta de `api/app/services/classifier.py`.
 *
 * Classificação heurística sem inventar dado: quando a evidência não
 * basta, o item vira `unclassified` — nunca um tipo adivinhado
 * (constitution, "IA e Classificação Nunca Inventam Dados").
 *
 * Função pura sobre entradas já interpretadas. Nesta feature só canais
 * chegam a ser gravados (D-006), mas a classificação continua
 * distinguindo os outros tipos: é assim que se sabe o que foi
 * **descartado de propósito**, e é isso que permite dizer "não é o
 * catálogo completo" com número real em vez de suposição.
 */

import type { ParsedEntry } from './m3uParser'

import type { CatalogItemKind } from './db'

export interface ClassifiedEntry {
  kind: CatalogItemKind
  name: string
  originalName: string
  group?: string
  url?: string
  seriesKey?: string
  seriesName?: string
  seriesId?: string
  seasonNumber?: number
  episodeNumber?: number
  streamExtension?: string
  providerStreamId?: string
  providerCategoryId?: string
  /** Capa declarada pela fonte (feature 015, D-002 do plan.md) — nunca para `kind: 'channel'` (FR-009). */
  iconUrl?: string
}

/**
 * Valida a **forma** de uma URL de capa — fonte única desta checagem
 * (feature 015, D-001b do plan.md), chamada tanto daqui (M3U, `tvg-logo`)
 * quanto de `xtreamConnector.ts` (`stream_icon`/`cover`), pra não
 * triplicar a mesma validação em cada captura. Ausente, vazia, ou que
 * `new URL(...)` rejeite vira `undefined` — nunca chega a `CatalogRecord`
 * como string inválida (FR-008). Nunca lança.
 */
export function normalizeIconUrl(raw: unknown): string | undefined {
  if (typeof raw !== 'string') return undefined
  const trimmed = raw.trim()
  if (trimmed === '') return undefined
  try {
    new URL(trimmed)
    return trimmed
  } catch {
    return undefined
  }
}

const EPISODE_PATTERN = /^(?<base>.*?)[\s._-]*S(?<season>\d{1,2})\s*E(?<episode>\d{1,3})\b.*$/i

const MOVIE_KEYWORDS = ['filme', 'filmes', 'movie', 'movies', 'vod']
const SERIES_KEYWORDS = ['série', 'séries', 'serie', 'series', 'seriado']
const CHANNEL_KEYWORDS = ['canal', 'canais', 'channel', 'channels', 'ao vivo', 'live', 'tv']

function groupKeywordKind(group: string | undefined): CatalogItemKind | undefined {
  if (!group) return undefined
  const lowered = group.toLowerCase()
  if (MOVIE_KEYWORDS.some((keyword) => lowered.includes(keyword))) return 'movie'
  if (SERIES_KEYWORDS.some((keyword) => lowered.includes(keyword))) return 'series'
  if (CHANNEL_KEYWORDS.some((keyword) => lowered.includes(keyword))) return 'channel'
  return undefined
}

export function classifyEntry(entry: ParsedEntry): ClassifiedEntry {
  // Feature 015 (D-001/FR-002/FR-008): nunca para canal (FR-009) — só
  // episódio (pra a série sintética herdar, D-003 de m3uSeriesGrouping.ts)
  // e filme/série capturam.
  const iconUrl = normalizeIconUrl(entry.attributes['tvg-logo'])

  const episodeMatch = EPISODE_PATTERN.exec(entry.name.trim())
  if (episodeMatch?.groups) {
    const base = episodeMatch.groups.base.replace(/[\s\-._]+$/, '').trim() || entry.name
    return {
      kind: 'episode',
      name: entry.name,
      originalName: entry.name,
      group: entry.group,
      url: entry.url,
      seriesKey: base.toLowerCase(),
      seriesName: base,
      seasonNumber: Number(episodeMatch.groups.season),
      episodeNumber: Number(episodeMatch.groups.episode),
      iconUrl,
    }
  }

  const byGroup = groupKeywordKind(entry.group)
  if (byGroup !== undefined) {
    return {
      kind: byGroup,
      name: entry.name,
      originalName: entry.name,
      group: entry.group,
      url: entry.url,
      iconUrl: byGroup === 'channel' ? undefined : iconUrl,
    }
  }

  // Evidência insuficiente — sem hierarquia nem tipo inventado.
  return {
    kind: 'unclassified',
    name: entry.name,
    originalName: entry.name,
    group: entry.group,
    url: entry.url,
  }
}
